std-bits/base/src/main/java/org/enso/base/parser/Separators.java

```diff
@@ -0,0 +1,131 @@
+  /** Check if the character is a separator. */
+  static boolean isSeparator(char c) {
+    return c == '.' || c == ',' || c == ' ' || c == '\'';
+  }
+
+  /**
+   * Strip out the specified separators and replace with just full stop for decimal. If any
+   * character other than a digit, thousands or decimal separator is encountered then return null.
+   * If multiple decimal separators are encountered then return null.
+   */
+  static CharSequence strip(
+      CharSequence value, int startIdx, int endIdx, char thousands, char decimal) {
+    boolean foundDecimal = false;
+    char[] results = new char[endIdx - startIdx];
+    int resultIdx = 0;
+    for (int i = startIdx; i < endIdx; i++) {
+      char c = value.charAt(i);
+      if (c == decimal) {
+        if (foundDecimal) {
+          return null;
+        }
+        results[resultIdx++] = '.';
+        foundDecimal = true;
+      } else if (isDigit(c)) {
+        results[resultIdx++] = c;
+      } else if (c != thousands) {
+        return null;
+      }
+    }
+    return CharBuffer.wrap(results, 0, resultIdx);
+  }
+
+  private static boolean validChar(char c, char first, char second) {
+    return isDigit(c)
+        || (first == NumberWithSeparators.Constants.NONE
+            ? isSeparator(c)
+            : (second == NumberWithSeparators.Constants.NONE
+                ? c == first || c == ',' || c == '.'
+                : c == second));
+  }
+
+  /**
+   * Find the number and separators section. Validate the spacing of separators. Return the
+   * separators found or null if invalid.
+   */
+  static Separators parse(CharSequence value, int idx, boolean integer) {
+    int endIdx = idx;
+    char firstSeparator = NumberWithSeparators.Constants.NONE;
+    char secondSeparator = NumberWithSeparators.Constants.NONE;
+
+    boolean firstWasSeparator = false;
+    int lastSeparator = -1;
+    int separatorCount = 0;
+
+    // Scan the text, find and validate spacing of separators.
+    // Space and ' are both valid thousands separators, but can't be second separator.
+    // ToDo: Cope with scientific notation.
+    char c = value.charAt(idx);
+    while (endIdx < value.length() && validChar(c, firstSeparator, secondSeparator)) {
+      if (!isDigit(c)) {
+        if (lastSeparator == -1) {
+          if (endIdx == idx) {
+            // If first digit is a separator then only valid if a decimal separator.
+            firstWasSeparator = true;
+            if (!integer && (c != '.' && c != ',')) {
+              return null;
+            }
+          }
+
+          firstSeparator = c;
+          separatorCount = 1;
+          lastSeparator = endIdx;
+        } else {
+          // Encountered a second separator -  must be 4 away from last separator.
+          if (endIdx != lastSeparator + 4) {
+            if (firstSeparator == ' ' && c == ' ') {
+              // Special case when encountered
+              break;
+            }
```

Comment:
What is special about this case? And why do we break from the loop then?

Reply:
Oh ok the special case is explained below.

Reply:
This is to deal with the delightful case of `1 234 567 USD`

the Symbol separated by the thousand delimiter.